File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/test-trellis-emit.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# test-trellis-emit.rb — unit test for TrellisEmit (the shared native-trellis
# emitter). Converter-agnostic, pure, no network. Canonical in shared/scripts;
# edit there, run tools/sync-shared.rb.
# Run: ruby scripts/test-trellis-emit.rb
require_relative 'lib/trellis_emit'

$fail = 0
def ok(name, cond); puts((cond ? '  ok  ' : 'FAIL  ') + name); $fail += 1 unless cond; end

def el(kind, cols = [{ 'id' => 'c-cat' }])
  { 'id' => 'el-1', 'kind' => kind, 'columns' => cols }
end

# ---- supported kinds → trellis set, right axis per orientation --------------
%w[bar-chart line-chart area-chart combo-chart scatter-chart donut-chart].each do |k|
  e = el(k)
  r = TrellisEmit.apply(e, facet_column_id: 'c-cat', orientation: :rows)
  ok("#{k}: rows → :trellised + trellis.rowsBy", r == :trellised &&
     e['trellis'] == { 'rowsBy' => [{ 'columnId' => 'c-cat' }] })

  e2 = el(k)
  r2 = TrellisEmit.apply(e2, facet_column_id: 'c-cat', orientation: :cols)
  ok("#{k}: cols → trellis.columnsBy (and NOT rowsBy)", r2 == :trellised &&
     e2['trellis'] == { 'columnsBy' => [{ 'columnId' => 'c-cat' }] })
end

# String orientation is accepted the same as the symbol.
e = el('bar-chart')
TrellisEmit.apply(e, facet_column_id: 'c-cat', orientation: 'rows')
ok('orientation may be a String', e['trellis'] == { 'rowsBy' => [{ 'columnId' => 'c-cat' }] })

# ---- pie-chart → donut + trellis (fallback) ---------------------------------
e = el('pie-chart')
r = TrellisEmit.apply(e, facet_column_id: 'c-cat', orientation: :cols)
ok('pie-chart → :fallback_donut', r == :fallback_donut)
ok('pie-chart kind flipped to donut-chart', e['kind'] == 'donut-chart')
ok('pie→donut carries the native trellis', e['trellis'] == { 'columnsBy' => [{ 'columnId' => 'c-cat' }] })
ok('pie-chart trellises? is true (converts)', TrellisEmit.trellises?('pie-chart'))

# ---- kpi-chart → needs-sibling signal, element untouched --------------------
e = el('kpi-chart')
r = TrellisEmit.apply(e, facet_column_id: 'c-cat', orientation: :cols)
ok('kpi-chart → :needs_sibling_fanout', r == :needs_sibling_fanout)
ok('kpi-chart left UNTOUCHED (no trellis, kind unchanged)',
   !e.key?('trellis') && e['kind'] == 'kpi-chart')
ok('kpi-chart trellises? is false', !TrellisEmit.trellises?('kpi-chart'))

# ---- pivot-table → needs-shelves signal, untouched --------------------------
e = el('pivot-table')
r = TrellisEmit.apply(e, facet_column_id: 'c-cat', orientation: :rows)
ok('pivot-table → :needs_pivot_shelves', r == :needs_pivot_shelves)
ok('pivot-table left UNTOUCHED', !e.key?('trellis') && e['kind'] == 'pivot-table')

# ---- table → flat no-op -----------------------------------------------------
e = el('table')
r = TrellisEmit.apply(e, facet_column_id: 'c-cat', orientation: :cols)
ok('table → :flat', r == :flat)
ok('table left UNTOUCHED', !e.key?('trellis'))

# ---- unknown kind → flat (default fallback) ---------------------------------
e = el('funnel-chart')
ok('unknown kind → :flat disposition', TrellisEmit.disposition('funnel-chart') == :flat)
r = TrellisEmit.apply(e, facet_column_id: 'c-cat', orientation: :cols)
ok('unknown kind → :flat + untouched', r == :flat && !e.key?('trellis'))

# ---- 2-D grid → BOTH keys, two different columns ----------------------------
e = el('bar-chart', [{ 'id' => 'c-row' }, { 'id' => 'c-col' }])
r = TrellisEmit.apply(e, facet_column_id: %w[c-row c-col], orientation: :grid)
ok('grid + two facet ids → :trellised', r == :trellised)
ok('grid → rowsBy AND columnsBy on the two distinct columns',
   e['trellis'] == { 'rowsBy' => [{ 'columnId' => 'c-row' }],
                     'columnsBy' => [{ 'columnId' => 'c-col' }] })

# A single-field grid wraps into a tile grid → columnsBy only (tableau parity).
e = el('bar-chart')
TrellisEmit.apply(e, facet_column_id: 'c-cat', orientation: :grid)
ok('grid + single facet id → columnsBy only (single-field tile grid)',
   e['trellis'] == { 'columnsBy' => [{ 'columnId' => 'c-cat' }] })

# ---- disposition is the single source of truth (non-mutating) ---------------
ok('disposition(bar-chart) == :trellised', TrellisEmit.disposition('bar-chart') == :trellised)
ok('disposition(pie-chart) == :fallback_donut', TrellisEmit.disposition('pie-chart') == :fallback_donut)
ok('disposition(kpi-chart) == :needs_sibling_fanout',
   TrellisEmit.disposition('kpi-chart') == :needs_sibling_fanout)
ok('disposition(pivot-table) == :needs_pivot_shelves',
   TrellisEmit.disposition('pivot-table') == :needs_pivot_shelves)
ok('disposition(table) == :flat', TrellisEmit.disposition('table') == :flat)
ok('SUPPORTED_KINDS is the documented readback-safe set',
   TrellisEmit::SUPPORTED_KINDS == %w[bar-chart line-chart area-chart combo-chart scatter-chart donut-chart])

puts
if $fail.zero?
  puts 'ALL PASS — TrellisEmit (supported→trellis by orientation; pie→donut; ' \
       'kpi/pivot/table fallbacks; 2-D grid → both keys)'
  exit 0
else
  puts "#{$fail} FAILURE(S)"
  exit 1
end

File: shared/scripts/test_trellis_emit.py
#!/usr/bin/env python3
"""test_trellis_emit.py — unit test for TrellisEmit (the shared native-trellis
emitter), the Python side. Byte-for-byte behavioural mirror of the Ruby
test-trellis-emit.rb: identical cases, identical expected shapes, results as the
mirrored STRINGS (Ruby returns symbols). Converter-agnostic, pure, no network.
Canonical in shared/scripts; edit there, run tools/sync-shared.rb.
Run: python3 scripts/test_trellis_emit.py
"""
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "lib"))
import trellis_emit as TE  # noqa: E402

_fail = 0


def ok(name, cond):
    global _fail
    print(("  ok  " if cond else "FAIL  ") + name)
    if not cond:
        _fail += 1


def el(kind, cols=None):
    return {"id": "el-1", "kind": kind, "columns": cols if cols is not None else [{"id": "c-cat"}]}


# ---- supported kinds -> trellis set, right axis per orientation --------------
for k in ("bar-chart", "line-chart", "area-chart", "combo-chart", "scatter-chart", "donut-chart"):
    e = el(k)
    r = TE.apply(e, "c-cat", "rows")
    ok("%s: rows -> trellised + trellis.rowsBy" % k,
       r == "trellised" and e["trellis"] == {"rowsBy": [{"columnId": "c-cat"}]})

    e2 = el(k)
    r2 = TE.apply(e2, "c-cat", "cols")
    ok("%s: cols -> trellis.columnsBy (and NOT rowsBy)" % k,
       r2 == "trellised" and e2["trellis"] == {"columnsBy": [{"columnId": "c-cat"}]})

# ---- pie-chart -> donut + trellis (fallback) --------------------------------
e = el("pie-chart")
r = TE.apply(e, "c-cat", "cols")
ok("pie-chart -> fallback_donut", r == "fallback_donut")
ok("pie-chart kind flipped to donut-chart", e["kind"] == "donut-chart")
ok("pie->donut carries the native trellis", e["trellis"] == {"columnsBy": [{"columnId": "c-cat"}]})
ok("pie-chart trellises() is true (converts)", TE.trellises("pie-chart"))

# ---- kpi-chart -> needs-sibling signal, element untouched -------------------
e = el("kpi-chart")
r = TE.apply(e, "c-cat", "cols")
ok("kpi-chart -> needs_sibling_fanout", r == "needs_sibling_fanout")
ok("kpi-chart left UNTOUCHED (no trellis, kind unchanged)",
   "trellis" not in e and e["kind"] == "kpi-chart")
ok("kpi-chart trellises() is false", not TE.trellises("kpi-chart"))

# ---- pivot-table -> needs-shelves signal, untouched -------------------------
e = el("pivot-table")
r = TE.apply(e, "c-cat", "rows")
ok("pivot-table -> needs_pivot_shelves", r == "needs_pivot_shelves")
ok("pivot-table left UNTOUCHED", "trellis" not in e and e["kind"] == "pivot-table")

# ---- table -> flat no-op ----------------------------------------------------
e = el("table")
r = TE.apply(e, "c-cat", "cols")
ok("table -> flat", r == "flat")
ok("table left UNTOUCHED", "trellis" not in e)

# ---- unknown kind -> flat (default fallback) --------------------------------
e = el("funnel-chart")
ok("unknown kind -> flat disposition", TE.disposition("funnel-chart") == "flat")
r = TE.apply(e, "c-cat", "cols")
ok("unknown kind -> flat + untouched", r == "flat" and "trellis" not in e)

# ---- 2-D grid -> BOTH keys, two different columns ---------------------------
e = el("bar-chart", [{"id": "c-row"}, {"id": "c-col"}])
r = TE.apply(e, ["c-row", "c-col"], "grid")
ok("grid + two facet ids -> trellised", r == "trellised")
ok("grid -> rowsBy AND columnsBy on the two distinct columns",
   e["trellis"] == {"rowsBy": [{"columnId": "c-row"}], "columnsBy": [{"columnId": "c-col"}]})

# A single-field grid wraps into a tile grid -> columnsBy only (tableau parity).
e = el("bar-chart")
TE.apply(e, "c-cat", "grid")
ok("grid + single facet id -> columnsBy only (single-field tile grid)",
   e["trellis"] == {"columnsBy": [{"columnId": "c-cat"}]})

# ---- disposition is the single source of truth (non-mutating) ---------------
ok("disposition(bar-chart) == trellised", TE.disposition("bar-chart") == "trellised")
ok("disposition(pie-chart) == fallback_donut", TE.disposition("pie-chart") == "fallback_donut")
ok("disposition(kpi-chart) == needs_sibling_fanout",
   TE.disposition("kpi-chart") == "needs_sibling_fanout")
ok("disposition(pivot-table) == needs_pivot_shelves",
   TE.disposition("pivot-table") == "needs_pivot_shelves")
ok("disposition(table) == flat", TE.disposition("table") == "flat")
ok("SUPPORTED_KINDS is the documented readback-safe set",
   list(TE.SUPPORTED_KINDS) == ["bar-chart", "line-chart", "area-chart",
                                "combo-chart", "scatter-chart", "donut-chart"])

print()
if _fail == 0:
    print("ALL PASS — TrellisEmit (supported->trellis by orientation; pie->donut; "
          "kpi/pivot/table fallbacks; 2-D grid -> both keys)")
    sys.exit(0)
else:
    print("%d FAILURE(S)" % _fail)
    sys.exit(1)
